msgpack/src/unpack.cpp

```diff
@@ -0,0 +1,171 @@
+				dst = list_unpacking(packed);
+			}
+			else
+			{
+				type_test();
+			}
+
+			return dst;
+		}
+
+		Value object_str(std::vector<Value> &args)
+		{
+			if(args.size() != 1)
+				throw std::runtime_error("ArgError : This function must have 1 argument");
+			if(args[0].valueType() != ValueType::String && args[0].valueType() != ValueType::List)
+				throw Ark::TypeError("The packed buffer must be a string or a list");
+			Value msg_object_str;
+			std::ostringstream str_buffer;
+			msgpack::object deserialized;
+
+			if(args[0].valueType() == ValueType::List)
+			{
+				list_unpacked_str(static_cast<Value>(args[0]).list(), str_buffer);
+				msg_object_str = Value(str_buffer.str());
+			}
+			else
+			{
+				std::string packed {static_cast<Value>(args[0]).string_ref()};
+				deserialized =  msgpack::unpack(packed.data(), packed.size()).get();
+				str_buffer << deserialized;
+				msg_object_str = Value(str_buffer.str());
+			}
+
+			return msg_object_str;
+		}
+
+		void list_unpacked_str(std::vector<Value> &buffer_list, std::ostringstream &stream)
+		{
+			msgpack::object deserialized;
+			stream << '[';
+
+			for(unsigned i {0}; i < buffer_list.size(); ++ i)
+			{
+				std::string buffer {static_cast<Value>(buffer_list[i]).string_ref()};
+				deserialized = msgpack::unpack(buffer.data(), buffer.size()).get();
+				if(i > 0)
+					stream << ' ';
+				stream << deserialized;
+			}
+
+			stream << ']';
+		}
+
+		Value list_unpacking(std::vector<Value> &buffer_list)
+		{
+			std::vector<Value> list;
+			bool ark_bool;
+			double ark_number;
+			std::string ark_string;
+			msgpack::object deserialized;
+			auto each_to_value = [&](void) {
+				try
+				{
+					deserialized.convert(ark_bool);
+					list.push_back(Value(ark_bool));
+				}
+				catch(const std::bad_cast &e)
+				{
+					try
+					{
+						deserialized.convert(ark_number);
+						list.push_back(Value(ark_number));
+					}
+					catch(const std::bad_cast &e)
+					{
+						try
+						{
+							deserialized.convert(ark_string);
+							list.push_back(Value(ark_string));
+						}
+						catch(std::exception &e) {}
+					}
+				}
+			};
+
+			for(unsigned i {0}; i < buffer_list.size(); ++ i)
+			{
+				std::string buffer {static_cast<Value>(buffer_list[i]).string_ref()};
+				deserialized = msgpack::unpack(buffer.data(), buffer.size()).get();
+				each_to_value();
+			}
+
+			return list;
+		}
+	}
+
+	Value unpack(std::vector<Value> &args)
+	{
+		if(args.size() != 1)
+			throw std::runtime_error("ArgError : This function must have 1 argument");
+		if(args[0].valueType() != ValueType::User && args[0].usertype().type_id() != std::type_index(typeid(msgpack::sbuffer)))
+			throw Ark::TypeError("The packed buffer must be msgpack::sbuffer");
+		msgpack::sbuffer* sbuf {static_cast<msgpack::sbuffer*>((args[0]).usertype().data())};
+		static msgpack::object_handle oh = msgpack::unpack(sbuf->data(), sbuf->size());
+
+		return Value(UserType(&oh));
```

Comment:
control functions

Reply:
Why ?

Reply:
usertypes need control functions to print them/destroy them

Reply:
Destroy ? No
But print perhaps :+1: :

Reply:
And in case of msgpack's objects printing with control functions seems hard

Reply:
Hard but not impossible

Why can't we delete msgpack objects?

Reply:
Explicit delete is not necessary for me
Let the VM delete all objects at end is better, I think

Reply:
What if an user has 10k objects?

Also, the delete function is called when an object goes out of scope and needs to be deleted, it's not just for explicit delete

Reply:
Reusabe objects ? Perhaps ? :thinking:

Reply:
the module shouldn't force the user to do something oddly specific to work correctly